File: lib/features/expenses/domain/expense_settlement.dart
import 'package:planzers/features/expenses/data/expense.dart';

/// Net balance per person in one currency (positive = others owe this person).
typedef BalancesByCurrency = Map<String, Map<String, double>>;

/// One suggested bank transfer / cash payment to settle debts.
class SuggestedTransfer {
  const SuggestedTransfer({
    required this.fromUserId,
    required this.toUserId,
    required this.amount,
    required this.currency,
  });

  final String fromUserId;
  final String toUserId;
  final double amount;
  final String currency;
}

/// Balances and transfer suggestions for one viewer within a single scope
/// (e.g. one expense post).
class ViewerSettlement {
  const ViewerSettlement({
    required this.balancesByCurrency,
    required this.suggestedTransfers,
  });

  /// Net balances from the provided expenses (caller scopes the list).
  final BalancesByCurrency balancesByCurrency;

  /// Transfers that involve the viewer as payer or payee; when there is no
  /// viewer id, every suggested transfer for that scope is returned.
  final List<SuggestedTransfer> suggestedTransfers;
}

/// Computes [ViewerSettlement] from [expenses] (already scoped, e.g. one post).
///
/// When [viewerUserId] is null or blank, every suggested transfer is returned.
ViewerSettlement computeViewerSettlement(
  Iterable<TripExpense> expenses,
  String? viewerUserId,
  {Iterable<SuggestedTransfer> settledTransfers = const []}
) {
  final balances = computeBalances(expenses);
  applySettledTransfersToBalances(
    balances: balances,
    settledTransfers: settledTransfers,
  );
  var transfers = suggestTransfers(balances);
  final v = viewerUserId?.trim();
  if (v != null && v.isNotEmpty) {
    transfers = transfers
        .where((t) => t.fromUserId == v || t.toUserId == v)
        .toList();
  }
  return ViewerSettlement(
    balancesByCurrency: balances,
    suggestedTransfers: transfers,
  );
}

/// Applies already-paid transfers on top of computed balances.
///
/// This makes paid transfers disappear from suggestions and impacts displayed
/// balances.
void applySettledTransfersToBalances({
  required BalancesByCurrency balances,
  required Iterable<SuggestedTransfer> settledTransfers,
}) {
  for (final transfer in settledTransfers) {
    final fromUserId = transfer.fromUserId.trim();
    final toUserId = transfer.toUserId.trim();
    final currency = transfer.currency.trim().toUpperCase();
    final amount = _roundMoney(transfer.amount);
    if (fromUserId.isEmpty || toUserId.isEmpty || currency.isEmpty) continue;
    if (amount <= _kBalanceEpsilon) continue;

    final bucket = balances.putIfAbsent(currency, () => <String, double>{});
    final fromBalance = (bucket[fromUserId] ?? 0) + amount;
    final toBalance = (bucket[toUserId] ?? 0) - amount;
    bucket[fromUserId] = _roundMoney(fromBalance);
    bucket[toUserId] = _roundMoney(toBalance);

    if (bucket[fromUserId]!.abs() <= _kBalanceEpsilon) {
      bucket.remove(fromUserId);
    }
    if (bucket[toUserId]!.abs() <= _kBalanceEpsilon) {
      bucket.remove(toUserId);
    }
    if (bucket.isEmpty) {
      balances.remove(currency);
    }
  }
}

/// Supported expense currencies for MVP (display + separate balance buckets).
const Set<String> kSupportedExpenseCurrencies = {'EUR', 'USD'};

/// Epsilon for floating-point comparisons (cent-level).
const double _kBalanceEpsilon = 0.009;

/// Computes per-user net balance by currency from shared expenses.
///
/// For each expense, each participant is debited their share (equal split or
/// explicit [TripExpense.participantShares]); [paidBy] is credited the full
/// amount paid (standard Splitwise / TriCount-style model).
BalancesByCurrency computeBalances(Iterable<TripExpense> expenses) {
  final result = <String, Map<String, double>>{};

  for (final expense in expenses) {
    final currency = expense.currency.trim().toUpperCase();
    if (currency.isEmpty) continue;

    final participants = expense.participantIds
        .map((id) => id.trim())
        .where((id) => id.isNotEmpty)
        .toList();
    if (participants.isEmpty) continue;

    final paidBy = expense.paidBy.trim();
    if (paidBy.isEmpty) continue;

    final amount = expense.amount;
    if (amount <= 0) continue;

    final bucket = result.putIfAbsent(currency, () => <String, double>{});

    final shares = _participantSharesForExpense(expense, participants);
    for (final uid in participants) {
      final share = shares[uid] ?? 0;
      bucket[uid] = (bucket[uid] ?? 0) - share;
    }
    bucket[paidBy] = (bucket[paidBy] ?? 0) + amount;
  }

  return result;
}

/// Resolves owed amount per participant; falls back to equal split when needed.
Map<String, double> _participantSharesForExpense(
  TripExpense expense,
  List<String> participants,
) {
  if (expense.splitMode != ExpenseSplitMode.customAmounts) {
    final n = participants.length;
    final per = n > 0 ? expense.amount / n : 0.0;
    return {for (final id in participants) id: per};
  }

  final raw = expense.participantShares;
  var sum = 0.0;
  final out = <String, double>{};
  for (final id in participants) {
    final v = raw[id];
    if (v == null || v < 0) {
      return {
        for (final uid in participants)
          uid: participants.isEmpty ? 0.0 : expense.amount / participants.length,
      };
    }
    out[id] = v;
    sum += v;
  }
  if ((sum - expense.amount).abs() > 0.02) {
    final n = participants.length;
    final per = n > 0 ? expense.amount / n : 0.0;
    return {for (final id in participants) id: per};
  }
  return out;
}

/// Greedy simplification: minimal number of transfers per currency to zero balances.
List<SuggestedTransfer> suggestTransfers(BalancesByCurrency balancesByCurrency) {
  final transfers = <SuggestedTransfer>[];

  for (final entry in balancesByCurrency.entries) {
    final currency = entry.key;
    final raw = entry.value;

    final working = <String, double>{};
    for (final e in raw.entries) {
      final v = _roundMoney(e.value);
      if (v.abs() > _kBalanceEpsilon) {
        working[e.key] = v;
      }
    }
    if (working.isEmpty) continue;

    _simplifyCurrency(working, currency, transfers);
  }

  return transfers;
}

void _simplifyCurrency(
  Map<String, double> balances,
  String currency,
  List<SuggestedTransfer> out,
) {
  while (true) {
    String? maxCreditor;
    double maxCredit = 0;
    String? maxDebtor;
    double maxDebt = 0;

    for (final e in balances.entries) {
      if (e.value > maxCredit) {
        maxCredit = e.value;
        maxCreditor = e.key;
      }
      if (e.value < maxDebt) {
        maxDebt = e.value;
        maxDebtor = e.key;
      }
    }

    if (maxCreditor == null ||
        maxDebtor == null ||
        maxCredit <= _kBalanceEpsilon ||
        maxDebt >= -_kBalanceEpsilon) {
      break;
    }

    final pay = _roundMoney(
      maxCredit < -maxDebt ? maxCredit : -maxDebt,
    );
    if (pay <= _kBalanceEpsilon) break;

    out.add(SuggestedTransfer(
      fromUserId: maxDebtor,
      toUserId: maxCreditor,
      amount: pay,
      currency: currency,
    ));

    balances[maxCreditor] = _roundMoney(maxCredit - pay);
    balances[maxDebtor] = _roundMoney(maxDebt + pay);

    if (balances[maxCreditor]!.abs() <= _kBalanceEpsilon) {
      balances.remove(maxCreditor);
    }
    if (balances[maxDebtor]!.abs() <= _kBalanceEpsilon) {
      balances.remove(maxDebtor);
    }
  }
}

double _roundMoney(double value) => (value * 100).round() / 100;

File: scripts/expense_settlement.js
'use strict';

/**
 * Utility scripts only (trip_expense_settlements.js, migrations, etc.).
 * Authoritative Cloud Functions copy: functions/expense_settlement.js (participantId naming).
 * When changing settlement algorithms, update both files.
 */

const BALANCE_EPSILON = 0.009;

function roundMoney(value) {
  return Math.round(value * 100) / 100;
}

function splitModeFromFirestore(raw) {
  const s = (raw == null ? '' : String(raw)).trim().toLowerCase();
  if (s === 'custom' || s === 'amounts' || s === 'montants') {
    return 'customAmounts';
  }
  return 'equal';
}

function participantSharesFromFirestore(raw) {
  if (raw == null || typeof raw !== 'object' || Array.isArray(raw)) {
    return {};
  }
  const out = {};
  for (const [key, value] of Object.entries(raw)) {
    const k = String(key).trim();
    if (!k) continue;
    const n = typeof value === 'number' ? value : Number(value);
    if (Number.isFinite(n)) out[k] = n;
  }
  return out;
}

/**
 * Resolves how many billing parts an id represents.
 *
 * @param {string} id
 * @param {Record<string, { parts: number }>} groupsMap  keyed by groupId
 * @returns {number}
 */
function resolveUnit(id, groupsMap) {
  const g = groupsMap[id];
  return g != null && g.parts > 0 ? g.parts : 1;
}

/** @typedef {{ id: string, groupId: string, amount: number, currency: string, paidBy: string, participantIds: string[], splitMode: string, participantShares: Record<string, number> }} TripExpense */

/**
 * @param {import('firebase-admin').firestore.DocumentSnapshot} doc
 * @returns {TripExpense}
 */
function tripExpenseFromDoc(doc) {
  const data = doc.data() || {};
  const amountRaw = data.amount;
  const amount =
    typeof amountRaw === 'number' && Number.isFinite(amountRaw)
      ? amountRaw
      : 0;

  return {
    id: doc.id,
    groupId: String(data.groupId || '').trim(),
    amount,
    currency: String(data.currency || 'EUR')
      .trim()
      .toUpperCase(),
    paidBy: String(data.paidBy || '').trim(),
    participantIds: ((data.participantIds || []))
      .map((e) => String(e).trim())
      .filter((id) => id.length > 0),
    splitMode: splitModeFromFirestore(data.splitMode),
    participantShares: participantSharesFromFirestore(data.participantShares),
  };
}

/**
 * @param {TripExpense} expense
 * @param {string[]} participants
 * @param {Record<string, { parts: number }>} [groupsMap]
 * @returns {Record<string, number>}
 */
function participantSharesForExpense(expense, participants, groupsMap = {}) {
  if (expense.splitMode !== 'customAmounts') {
    const totalParts = participants.reduce(
      (sum, id) => sum + resolveUnit(id, groupsMap),
      0
    );
    if (totalParts <= 0) {
      return Object.fromEntries(participants.map((id) => [id, 0]));
    }
    return Object.fromEntries(
      participants.map((id) => [
        id,
        expense.amount * resolveUnit(id, groupsMap) / totalParts,
      ])
    );
  }

  const raw = expense.participantShares;
  let sum = 0;
  const out = {};
  for (const id of participants) {
    const v = raw[id];
    if (v == null || v < 0) {
      const totalParts = participants.reduce(
        (s, i) => s + resolveUnit(i, groupsMap),
        0
      );
      if (totalParts <= 0) {
        return Object.fromEntries(participants.map((uid) => [uid, 0]));
      }
      return Object.fromEntries(
        participants.map((uid) => [
          uid,
          expense.amount * resolveUnit(uid, groupsMap) / totalParts,
        ])
      );
    }
    out[id] = v;
    sum += v;
  }
  if (Math.abs(sum - expense.amount) > 0.02) {
    const totalParts = participants.reduce(
      (s, i) => s + resolveUnit(i, groupsMap),
      0
    );
    if (totalParts <= 0) {
      return Object.fromEntries(participants.map((id) => [id, 0]));
    }
    return Object.fromEntries(
      participants.map((id) => [
        id,
        expense.amount * resolveUnit(id, groupsMap) / totalParts,
      ])
    );
  }
  return out;
}

/**
 * @param {Iterable<TripExpense>} expenses
 * @param {Record<string, { parts: number }>} [groupsMap]  keyed by groupId
 * @returns {Record<string, Record<string, number>>}
 */
function computeBalances(expenses, groupsMap = {}) {
  /** @type {Record<string, Record<string, number>>} */
  const result = {};

  for (const expense of expenses) {
    const currency = expense.currency.trim().toUpperCase();
    if (!currency) continue;

    const participants = expense.participantIds
      .map((id) => id.trim())
      .filter((id) => id.length > 0);
    if (participants.length === 0) continue;

    const paidBy = expense.paidBy.trim();
    if (!paidBy) continue;

    const amount = expense.amount;
    if (amount <= 0) continue;

    if (!result[currency]) result[currency] = {};
    const bucket = result[currency];

    const shares = participantSharesForExpense(expense, participants, groupsMap);
    for (const uid of participants) {
      const share = shares[uid] ?? 0;
      bucket[uid] = (bucket[uid] ?? 0) - share;
    }
    bucket[paidBy] = (bucket[paidBy] ?? 0) + amount;
  }

  return result;
}

/**
 * @typedef {{ fromUserId: string, toUserId: string, amount: number, currency: string }} SuggestedTransfer
 */

/**
 * @param {Record<string, Record<string, number>>} balances
 * @param {Iterable<SuggestedTransfer>} settledTransfers
 */
function applySettledTransfersToBalances(balances, settledTransfers) {
  for (const transfer of settledTransfers) {
    const fromUserId = String(transfer.fromUserId || '').trim();
    const toUserId = String(transfer.toUserId || '').trim();
    const currency = String(transfer.currency || '')
      .trim()
      .toUpperCase();
    const amount = roundMoney(transfer.amount);
    if (!fromUserId || !toUserId || !currency) continue;
    if (amount <= BALANCE_EPSILON) continue;

    if (!balances[currency]) balances[currency] = {};
    const bucket = balances[currency];
    const fromBalance = (bucket[fromUserId] ?? 0) + amount;
    const toBalance = (bucket[toUserId] ?? 0) - amount;
    bucket[fromUserId] = roundMoney(fromBalance);
    bucket[toUserId] = roundMoney(toBalance);

    if (Math.abs(bucket[fromUserId]) <= BALANCE_EPSILON) {
      bucket[fromUserId] = 0;
    }
    if (Math.abs(bucket[toUserId]) <= BALANCE_EPSILON) {
      bucket[toUserId] = 0;
    }
  }
}

/**
 * @param {Record<string, number>} balances
 * @param {string} currency
 * @param {SuggestedTransfer[]} out
 */
function simplifyCurrency(balances, currency, out) {
  while (true) {
    let maxCreditor = null;
    let maxCredit = 0;
    let maxDebtor = null;
    let maxDebt = 0;

    for (const [id, value] of Object.entries(balances)) {
      if (value > maxCredit) {
        maxCredit = value;
        maxCreditor = id;
      }
      if (value < maxDebt) {
        maxDebt = value;
        maxDebtor = id;
      }
    }

    if (
      maxCreditor == null ||
      maxDebtor == null ||
      maxCredit <= BALANCE_EPSILON ||
      maxDebt >= -BALANCE_EPSILON
    ) {
      break;
    }

    const pay = roundMoney(maxCredit < -maxDebt ? maxCredit : -maxDebt);
    if (pay <= BALANCE_EPSILON) break;

    out.push({
      fromUserId: maxDebtor,
      toUserId: maxCreditor,
      amount: pay,
      currency,
    });

    balances[maxCreditor] = roundMoney(maxCredit - pay);
    balances[maxDebtor] = roundMoney(maxDebt + pay);

    if (Math.abs(balances[maxCreditor]) <= BALANCE_EPSILON) {
      delete balances[maxCreditor];
    }
    if (Math.abs(balances[maxDebtor]) <= BALANCE_EPSILON) {
      delete balances[maxDebtor];
    }
  }
}

/**
 * @param {Record<string, Record<string, number>>} balancesByCurrency
 * @returns {SuggestedTransfer[]}
 */
function suggestTransfers(balancesByCurrency) {
  /** @type {SuggestedTransfer[]} */
  const transfers = [];

  for (const [currency, raw] of Object.entries(balancesByCurrency)) {
    /** @type {Record<string, number>} */
    const working = {};
    for (const [id, value] of Object.entries(raw)) {
      const v = roundMoney(value);
      if (Math.abs(v) > BALANCE_EPSILON) {
        working[id] = v;
      }
    }
    if (Object.keys(working).length === 0) continue;
    simplifyCurrency(working, currency, transfers);
  }

  return transfers;
}

/**
 * @param {Iterable<TripExpense>} expenses
 * @param {Iterable<SuggestedTransfer>} [settledTransfers]
 * @param {Record<string, { parts: number }>} [groupsMap]  keyed by groupId
 * @returns {{ balancesByCurrency: Record<string, Record<string, number>>, suggestedTransfers: SuggestedTransfer[] }}
 */
function computeSettlement(expenses, settledTransfers = [], groupsMap = {}) {
  const balances = computeBalances(expenses, groupsMap);
  applySettledTransfersToBalances(balances, settledTransfers);
  const suggestedTransfers = suggestTransfers(balances);
  return { balancesByCurrency: balances, suggestedTransfers };
}

module.exports = {
  BALANCE_EPSILON,
  roundMoney,
  resolveUnit,
  tripExpenseFromDoc,
  computeBalances,
  applySettledTransfersToBalances,
  suggestTransfers,
  computeSettlement,
};
